// payload/dist/globals/config/schema.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const joi_1 = __importDefault(require("joi"));
const componentSchema_1 = require("../../utilities/componentSchema");
const globalSchema = joi_1.default.object().keys({
    slug: joi_1.default.string().required(),
    label: joi_1.default.string(),
    admin: joi_1.default.object({
        description: joi_1.default.alternatives().try(joi_1.default.string(), componentSchema_1.componentSchema),
    }),
    hooks: joi_1.default.object({
        beforeValidate: joi_1.default.array().items(joi_1.default.func()),
        beforeChange: joi_1.default.array().items(joi_1.default.func()),
        afterChange: joi_1.default.array().items(joi_1.default.func()),
        beforeRead: joi_1.default.array().items(joi_1.default.func()),
        afterRead: joi_1.default.array().items(joi_1.default.func()),
    }),
    access: joi_1.default.object({
        read: joi_1.default.func(),
        update: joi_1.default.func(),
    }),
    fields: joi_1.default.array(),
}).unknown();
exports.default = globalSchema;
