// payload/dist/auth/operations/registerFirstUser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const errors_1 = require("../../errors");
async function registerFirstUser(args) {
    const { collection: { Model, config: { slug, auth: { verify, }, }, }, } = args;
    const count = await Model.countDocuments({});
    if (count >= 1)
        throw new errors_1.Forbidden();
    // /////////////////////////////////////
    // Register first user
    // /////////////////////////////////////
    let result = await this.operations.collections.create({
        ...args,
        overrideAccess: true,
    });
    // auto-verify (if applicable)
    if (verify) {
        await this.update({
            id: result.id,
            collection: slug,
            data: {
                _verified: true,
            },
        });
    }
    // /////////////////////////////////////
    // Log in new user
    // /////////////////////////////////////
    const { token } = await this.operations.collections.auth.login({
        ...args,
    });
    result = {
        ...result,
        token,
    };
    return {
        message: 'Registered and logged in successfully. Welcome!',
        user: result,
    };
}
exports.default = registerFirstUser;
