// @faceless-ui/scroll-info/dist/ScrollInfoProvider/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importStar(require("react"));
var ScrollInfoContext_1 = __importDefault(require("../ScrollInfoContext"));
var ScrollInfoProvider = /** @class */ (function (_super) {
    __extends(ScrollInfoProvider, _super);
    function ScrollInfoProvider(props) {
        var _this = _super.call(this, props) || this;
        _this.requestAnimation = function (e) {
            var animationScheduled = _this.state.animationScheduled;
            if (!animationScheduled) {
                _this.setState({
                    animationScheduled: true,
                }, function () { return requestAnimationFrame(function (timestamp) { return _this.updateScrollInfo(e, timestamp); }); });
            }
        };
        _this.updateScrollInfo = function (e, timestamp) {
            var _a = _this.state, prevScrollX = _a.x, prevScrollY = _a.y, prevXDirection = _a.xDirection, prevYDirection = _a.yDirection, prevEventsFired = _a.eventsFired, prevHasScrolled = _a.hasScrolled;
            // Set currentScroll to zero on mount for cross-browser compatibility --
            // Some browsers populate the cached window.pageOffset at different points of the component lifecycle.
            // Chrome mounts with cached window.pageOffset coordinates even before firing the respective cached scroll event.
            // Neither Safari and FireFox populate these coordinates until this cached scroll (natively triggered by the browser).
            // The presence of a timestamp indicates that the caller of this method was not componentDidMount,
            // but rather a true scroll event via requestAnimationFrame. Keep at zero otherwise.
            var hasScrolled = prevHasScrolled || Boolean(timestamp);
            var currentScrollX = hasScrolled ? window.pageXOffset : 0;
            var currentScrollY = hasScrolled ? window.pageYOffset : 0;
            // Only increment the eventsFired state after the first true scroll event.
            var eventsFired = hasScrolled ? prevEventsFired + 1 : prevEventsFired;
            var xDifference = currentScrollX - prevScrollX;
            var yDifference = currentScrollY - prevScrollY;
            var xPercentage = (currentScrollX / (document.body.scrollWidth - window.innerWidth)) * 100;
            var yPercentage = (currentScrollY / (document.body.scrollHeight - window.innerHeight)) * 100;
            var totalPercentage = (xPercentage + yPercentage) / 2;
            var xDirection = xDifference > 0 ? 'right' : xDifference < 0 ? 'left' : prevXDirection; // eslint-disable-line no-nested-ternary
            var yDirection = yDifference > 0 ? 'down' : yDifference < 0 ? 'up' : prevYDirection; // eslint-disable-line no-nested-ternary
            _this.setState({
                x: currentScrollX,
                y: currentScrollY,
                xDifference: xDifference,
                yDifference: yDifference,
                xDirection: xDirection,
                yDirection: yDirection,
                xPercentage: xPercentage,
                yPercentage: yPercentage,
                totalPercentage: totalPercentage,
                eventsFired: eventsFired,
                hasScrolled: hasScrolled,
                animationScheduled: false,
            });
        };
        _this.state = {
            x: 0,
            y: 0,
            xDifference: 0,
            yDifference: 0,
            xDirection: undefined,
            yDirection: undefined,
            xPercentage: 0,
            yPercentage: 0,
            totalPercentage: 0,
            eventsFired: 0,
            hasScrolled: false,
            animationScheduled: false,
        };
        return _this;
    }
    ScrollInfoProvider.prototype.componentDidMount = function () {
        window.addEventListener('scroll', this.requestAnimation);
        this.updateScrollInfo();
    };
    ScrollInfoProvider.prototype.componentWillUnmount = function () {
        window.removeEventListener('scroll', this.requestAnimation);
    };
    ScrollInfoProvider.prototype.render = function () {
        var children = this.props.children;
        var scrollInfo = __assign({}, this.state);
        delete scrollInfo.hasScrolled;
        delete scrollInfo.animationScheduled;
        return (react_1.default.createElement(ScrollInfoContext_1.default.Provider, { value: __assign({}, scrollInfo) }, children && children));
    };
    return ScrollInfoProvider;
}(react_1.Component));
exports.default = ScrollInfoProvider;
