// payload/dist/collections/init.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const mongoose_1 = __importDefault(require("mongoose"));
const express_1 = __importDefault(require("express"));
const passport_1 = __importDefault(require("passport"));
const passport_local_mongoose_1 = __importDefault(require("passport-local-mongoose"));
const passport_local_1 = __importDefault(require("passport-local"));
const apiKey_1 = __importDefault(require("../auth/strategies/apiKey"));
const buildSchema_1 = __importDefault(require("./buildSchema"));
const bindCollection_1 = __importDefault(require("./bindCollection"));
const LocalStrategy = passport_local_1.default.Strategy;
function registerCollections(ctx) {
    ctx.config.collections = ctx.config.collections.map((collection) => {
        const formattedCollection = collection;
        const schema = buildSchema_1.default(formattedCollection, ctx.config);
        if (collection.auth) {
            schema.plugin(passport_local_mongoose_1.default, {
                usernameField: 'email',
            });
            const { maxLoginAttempts, lockTime } = collection.auth;
            if (maxLoginAttempts > 0) {
                // eslint-disable-next-line func-names
                schema.methods.incLoginAttempts = function (cb) {
                    // Expired lock, restart count at 1
                    if (this.lockUntil && this.lockUntil < Date.now()) {
                        return this.updateOne({
                            $set: { loginAttempts: 1 },
                            $unset: { lockUntil: 1 },
                        }, cb);
                    }
                    const updates = { $inc: { loginAttempts: 1 } };
                    // Lock the account if at max attempts and not already locked
                    if (this.loginAttempts + 1 >= maxLoginAttempts && !this.isLocked) {
                        updates.$set = { lockUntil: Date.now() + lockTime };
                    }
                    return this.updateOne(updates, cb);
                };
                // eslint-disable-next-line func-names
                schema.methods.resetLoginAttempts = function (cb) {
                    return this.updateOne({
                        $set: { loginAttempts: 0 },
                        $unset: { lockUntil: 1 },
                    }, cb);
                };
            }
        }
        ctx.collections[formattedCollection.slug] = {
            Model: mongoose_1.default.model(formattedCollection.slug, schema),
            config: formattedCollection,
        };
        // If not local, open routes
        if (!ctx.local) {
            const router = express_1.default.Router();
            const { slug } = collection;
            router.all(`/${slug}*`, bindCollection_1.default(ctx.collections[formattedCollection.slug]));
            const { create, find, update, findByID, delete: deleteHandler, } = ctx.requestHandlers.collections;
            if (collection.auth) {
                const AuthCollection = ctx.collections[formattedCollection.slug];
                passport_1.default.use(new LocalStrategy(AuthCollection.Model.authenticate()));
                if (collection.auth.useAPIKey) {
                    passport_1.default.use(`${AuthCollection.config.slug}-api-key`, apiKey_1.default(ctx, AuthCollection));
                }
                const { init, login, logout, refresh, me, registerFirstUser, forgotPassword, resetPassword, verifyEmail, unlock, } = ctx.requestHandlers.collections.auth;
                if (collection.auth.verify) {
                    router
                        .route(`/${slug}/verify/:token`)
                        .post(verifyEmail);
                }
                if (collection.auth.maxLoginAttempts > 0) {
                    router
                        .route(`/${slug}/unlock`)
                        .post(unlock);
                }
                router
                    .route(`/${slug}/init`)
                    .get(init);
                router
                    .route(`/${slug}/login`)
                    .post(login);
                router
                    .route(`/${slug}/logout`)
                    .post(logout);
                router
                    .route(`/${slug}/refresh-token`)
                    .post(refresh);
                router
                    .route(`/${slug}/me`)
                    .get(me);
                router
                    .route(`/${slug}/first-register`)
                    .post(registerFirstUser);
                router
                    .route(`/${slug}/forgot-password`)
                    .post(forgotPassword);
                router
                    .route(`/${slug}/reset-password`)
                    .post(resetPassword);
            }
            router.route(`/${slug}`)
                .get(find)
                .post(create);
            router.route(`/${slug}/:id`)
                .put(update)
                .get(findByID)
                .delete(deleteHandler);
            ctx.router.use(router);
        }
        return formattedCollection;
    });
}
exports.default = registerCollections;
