// payload/dist/config/load.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
/* eslint-disable import/no-dynamic-require */
/* eslint-disable global-require */
const path_1 = __importDefault(require("path"));
const find_1 = __importDefault(require("./find"));
const validate_1 = __importDefault(require("./validate"));
const babel_config_1 = __importDefault(require("../babel.config"));
const removedExtensions = ['.scss', '.css', '.svg', '.png', '.jpg', '.eot', '.ttf', '.woff', '.woff2'];
const loadConfig = () => {
    const configPath = find_1.default();
    removedExtensions.forEach((ext) => {
        require.extensions[ext] = () => null;
    });
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    require('@babel/register')({
        ...babel_config_1.default,
        extensions: ['.ts', '.tsx', '.js', '.jsx'],
        env: {
            development: {
                sourceMaps: 'inline',
                retainLines: true,
            },
        },
        ignore: [
            /node_modules[\\/](?!payload[\\/]dist[\\/]admin|payload[\\/]components).*/,
        ],
    });
    // eslint-disable-next-line @typescript-eslint/no-var-requires
    let config = require(configPath);
    if (config.default)
        config = config.default;
    const validatedConfig = validate_1.default(config);
    return {
        ...validatedConfig,
        paths: {
            ...(validatedConfig.paths || {}),
            configDir: path_1.default.dirname(configPath),
            config: configPath,
        },
    };
};
exports.default = loadConfig;
