// payload/dist/mongoose/connect.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const mongoose_1 = __importDefault(require("mongoose"));
const logger_1 = __importDefault(require("../utilities/logger"));
const testCredentials_1 = require("./testCredentials");
const logger = logger_1.default();
const connectMongoose = async (url, options, local) => {
    let urlToConnect = url;
    let successfulConnectionMessage = 'Connected to Mongo server successfully!';
    const connectionOptions = {
        ...options,
        useNewUrlParser: true,
        useUnifiedTopology: true,
        useCreateIndex: true,
        autoIndex: true,
        useFindAndModify: false,
    };
    if (process.env.NODE_ENV === 'test') {
        if (local) {
            urlToConnect = `${testCredentials_1.connection.url}:${testCredentials_1.connection.port}/${testCredentials_1.connection.name}`;
        }
        else {
            connectionOptions.dbName = 'payloadmemory';
            // eslint-disable-next-line global-require, @typescript-eslint/no-var-requires
            const { MongoMemoryServer } = require('mongodb-memory-server');
            const mongo = await MongoMemoryServer.create({
                instance: {
                    dbName: testCredentials_1.connection.name,
                    port: testCredentials_1.connection.port,
                },
            });
            urlToConnect = mongo.getUri();
            successfulConnectionMessage = 'Connected to in-memory Mongo server successfully!';
        }
    }
    try {
        await mongoose_1.default.connect(urlToConnect, connectionOptions);
        logger.info(successfulConnectionMessage);
    }
    catch (err) {
        logger.error(`Error: cannot connect to MongoDB. Details: ${err.message}`, err);
        process.exit(1);
    }
};
exports.default = connectMongoose;
