// payload/dist/fields/relationshipPopulationPromise.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const executeAccess_1 = __importDefault(require("../auth/executeAccess"));
const types_1 = require("./config/types");
const populate = async ({ depth, currentDepth, req, overrideAccess, dataReference, data, field, index, id, payload, }) => {
    const dataToUpdate = dataReference;
    const fieldAsRelationship = field;
    const relation = Array.isArray(fieldAsRelationship.relationTo) ? data.relationTo : fieldAsRelationship.relationTo;
    const relatedCollection = payload.collections[relation];
    if (relatedCollection) {
        const accessResult = !overrideAccess ? await executeAccess_1.default({ req, disableErrors: true, id }, relatedCollection.config.access.read) : true;
        let populatedRelationship = null;
        if (accessResult && (depth && currentDepth <= depth)) {
            let idString = Array.isArray(fieldAsRelationship.relationTo) ? data.value : data;
            if (typeof idString !== 'string' && typeof (idString === null || idString === void 0 ? void 0 : idString.toString) === 'function') {
                idString = idString.toString();
            }
            populatedRelationship = await payload.operations.collections.findByID({
                req,
                collection: relatedCollection,
                id: idString,
                currentDepth: currentDepth + 1,
                overrideAccess,
                disableErrors: true,
                depth,
            });
        }
        // If access control fails, update value to null
        // If populatedRelationship comes back, update value
        if (!accessResult || populatedRelationship) {
            if (typeof index === 'number') {
                if (Array.isArray(fieldAsRelationship.relationTo)) {
                    dataToUpdate[field.name][index].value = populatedRelationship;
                }
                else {
                    dataToUpdate[field.name][index] = populatedRelationship;
                }
            }
            else if (Array.isArray(fieldAsRelationship.relationTo)) {
                dataToUpdate[field.name].value = populatedRelationship;
            }
            else {
                dataToUpdate[field.name] = populatedRelationship;
            }
        }
    }
};
const relationshipPopulationPromise = ({ data, field, depth, currentDepth, req, overrideAccess, payload, }) => async () => {
    const resultingData = data;
    const populateDepth = types_1.fieldHasMaxDepth(field) && field.maxDepth < depth ? field.maxDepth : depth;
    if (types_1.fieldSupportsMany(field) && field.hasMany && Array.isArray(data[field.name])) {
        const rowPromises = [];
        data[field.name].forEach((relatedDoc, index) => {
            const rowPromise = async () => {
                if (relatedDoc) {
                    await populate({
                        depth: populateDepth,
                        currentDepth,
                        req,
                        overrideAccess,
                        data: relatedDoc,
                        dataReference: resultingData,
                        field,
                        index,
                        payload,
                    });
                }
            };
            rowPromises.push(rowPromise());
        });
        await Promise.all(rowPromises);
    }
    else if (data[field.name]) {
        await populate({
            depth: populateDepth,
            currentDepth,
            req,
            overrideAccess,
            dataReference: resultingData,
            data: data[field.name],
            field,
            payload,
        });
    }
};
exports.default = relationshipPopulationPromise;
