// payload/dist/fields/richTextRelationshipPromise.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const populate = async ({ id, collection, data, overrideAccess, depth, currentDepth, payload, req, }) => {
    const dataRef = data;
    const doc = await payload.operations.collections.findByID({
        req: {
            ...req,
            payloadAPI: 'local',
        },
        collection,
        id,
        currentDepth: currentDepth + 1,
        overrideAccess,
        disableErrors: true,
        depth,
    });
    if (doc) {
        dataRef.value = doc;
    }
    else {
        dataRef.value = null;
    }
};
const recurseRichText = ({ req, children, payload, overrideAccess = false, depth, currentDepth = 0, field, promises, }) => {
    if (Array.isArray(children)) {
        children.forEach((element) => {
            var _a;
            const collection = payload.collections[element === null || element === void 0 ? void 0 : element.relationTo];
            if (element.type === 'relationship'
                && ((_a = element === null || element === void 0 ? void 0 : element.value) === null || _a === void 0 ? void 0 : _a.id)
                && collection
                && (depth && currentDepth <= depth)) {
                promises.push(populate({
                    req,
                    id: element.value.id,
                    data: element,
                    overrideAccess,
                    depth,
                    currentDepth,
                    payload,
                    field,
                    collection,
                }));
            }
            if (element === null || element === void 0 ? void 0 : element.children) {
                recurseRichText({
                    req,
                    children: element.children,
                    payload,
                    overrideAccess,
                    depth,
                    currentDepth,
                    field,
                    promises,
                });
            }
        });
    }
};
const richTextRelationshipPromise = ({ req, data, payload, overrideAccess, depth, currentDepth, field, }) => async () => {
    const promises = [];
    recurseRichText({
        req,
        children: data[field.name],
        payload,
        overrideAccess,
        depth,
        currentDepth,
        field,
        promises,
    });
    await Promise.all(promises);
};
exports.default = richTextRelationshipPromise;
